consensus/e2e_tests/utils_test.go

```diff
@@ -673,68 +700,78 @@ func WaitForNodeToSync(
 	unsyncedNode *shared.Node,
 	allNodes IdToNodeMapping,
 	targetHeight uint64,
-) error {
+) {
 	currentHeight := unsyncedNode.GetBus().GetConsensusModule().CurrentHeight()
 
-	for i := currentHeight; i <= targetHeight; i++ {
-		blockRequest, err := waitForNodeToRequestMissingBlock(t, clck, eventsChannel, allNodes, currentHeight, targetHeight)
-		if err != nil {
-			return err
-		}
+	for currentHeight < targetHeight {
+		// waiting for unsynced node to request missing block
+		blockRequests, err := waitForNodeToRequestMissingBlock(t, clck, eventsChannel)
+		require.NoError(t, err)
 
-		blockResponse, err := waitForNodeToReceiveMissingBlock(t, clck, eventsChannel, allNodes, blockRequest)
-		if err != nil {
-			return err
-		}
+		// broadcast one of the requests to all nodes
+		P2PBroadcast(t, allNodes, blockRequests[0])
+		advanceTime(t, clck, 10*time.Millisecond)
 
-		err = waitForNodeToCatchUp(t, clck, eventsChannel, unsyncedNode, blockResponse, targetHeight)
-		if err != nil {
-			return err
-		}
+		// wait to receive replies from all nodes
+		blockResponses, err := waitForNodesToReplyToBlockRequest(t, clck, eventsChannel)
+		require.NoError(t, err)
+
+		// send one of the block responses to the unsynced node
+		P2PSend(t, unsyncedNode, blockResponses[0])
+		advanceTime(t, clck, 10*time.Millisecond)
+
+		// waiting for node to reach to the next height (currentHeight + 1)
+		err = waitForNodeToCatchUp(t, clck, eventsChannel, unsyncedNode, currentHeight+1)
+		require.NoError(t, err)
+
+		currentHeight = unsyncedNode.GetBus().GetConsensusModule().CurrentHeight()
 	}
-	return nil
 }
 
-// TODO(#352): implement this function.
 // waitForNodeToRequestMissingBlock waits for unsynced node to request missing block form the network
 func waitForNodeToRequestMissingBlock(
 	t *testing.T,
 	clck *clock.Mock,
 	eventsChannel modules.EventsChannel,
-	allNodes IdToNodeMapping,
-	startingHeight uint64,
-	targetHeight uint64,
-) (*anypb.Any, error) {
+) ([]*anypb.Any, error) {
 
-	return &anypb.Any{}, nil
+	errMsg := "Error waiting for StateSync Block Request Messages"
+	msgs, err := WaitForNetworkStateSyncEvents(t, clck, eventsChannel, errMsg, numValidators, 250, true)
+	require.NoError(t, err)
+
+	return msgs, err
 }
 
-// TODO(#352): implement this function.
-// waitForNodeToReceiveMissingBlock requests block request of the unsynced node
-// for given node to node to catch up to the target height by sending the requested block.
-func waitForNodeToReceiveMissingBlock(
+// waitForNodesToReplyToBlockRequest waits for nodes to send back requested block
+func waitForNodesToReplyToBlockRequest(
```

Comment:
The business logic seems to be the same between this function and the one above it. What am I missing?